internal/rpc/digest_md5_integrity.go

```diff
@@ -0,0 +1,121 @@
+package rpc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"syscall"
+	"time"
+)
+
+type digestMD5IntegrityConn struct {
+	baseConnWrapper
+	buf bytes.Buffer
+
+	kic []byte
+	kis []byte
+
+	sendSeqNum int
+	readSeqNum int
+}
+
+func newDigestMD5IntegrityConn(conn net.Conn, kic, kis []byte) digestMD5Handler {
+	return &digestMD5IntegrityConn{
+		baseConnWrapper: baseConnWrapper{conn: conn},
+		kic:             kic,
+		kis:             kis,
+	}
+}
+
+func (d *digestMD5IntegrityConn) Write(b []byte) (n int, err error) {
+	if !d.writeDeadline.IsZero() && d.writeDeadline.Before(time.Now()) {
+		return 0, syscall.ETIMEDOUT
+	}
+
+	inputLen := len(b)
+	var seqBuf [4]byte
+	lenEncodeBytes(d.sendSeqNum, seqBuf[:])
+
+	outputLen := macDataLen + inputLen + macHMACLen + macMsgTypeLen + macSeqNumLen
+
+	n = 4
+	err = binary.Write(d.conn, binary.BigEndian, int32(outputLen-macDataLen))
```

Comment:
Too many calls to `conn.Write` here - we should try to just do one. Better to write to a buffer and then `io.Copy` it out.

Reply:
(It's not really about hyper-optimizing, but we can afford to be hygienic here)

Reply:
I originally had that but when i was using a buffer that I reused, i ended up with a data race here in the test that tests `block_write_stream`. By having it write to the connection directly we avoid the race condition because there's no intermediate buffer. I really didn't want to try tracing down the race condition that was happening when i reused a write buffer :(

Reply:
Hm, how could there be a race? The buffer is only used for the duration of this call.

Reply:
I had put the buffer onto the struct itself like you suggested because you didn't want the allocations in here, so the buffer would be shared between write calls (look at the write implementation in the privacy connection for an example)

Reply:
Hmm, I think it might have been me sharing the buffer between the read and writes rather than having a readbuffer and writebuffer. Looks like i got it all squared away now using a buffer here to write to and then using `buffer.WriteTo` to actually write it all to the connection. Looks good to me. let me know what you think.